pkg/controller/scyllacluster/sync_statefulsets.go

```diff
@@ -0,0 +1,728 @@
+				return
+			}
+
+			for _, keyspace := range keyspaces {
+				err := scyllaClient.TakeSnapshot(ctx, host, snapshotTag, keyspace)
+				if err != nil {
+					errCh <- fmt.Errorf("can't take a snapshot on host %q and keyspace %q: %w", host, keyspace, err)
+					return
+				}
+			}
+		}()
+	}
+	wg.Wait()
+	close(errCh)
+	var errs []error
+	for err := range errCh {
+		errs = append(errs, err)
+	}
+	err := utilerrors.NewAggregate(errs)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// beforeUpgrade runs hooks before a cluster upgrade starts.
+// It returns true if the action is done, false if the caller should repeat later.
+func (scc *Controller) beforeUpgrade(ctx context.Context, sc *scyllav1.ScyllaCluster, services map[string]*corev1.Service) (bool, error) {
+	klog.V(2).InfoS("Running pre-upgrade hook", "ScyllaCluster", klog.KObj(sc))
+	defer klog.V(2).InfoS("Finished running pre-upgrade hook", "ScyllaCluster", klog.KObj(sc))
+
+	hosts, err := controllerhelpers.GetRequiredScyllaHosts(sc, services)
+	if err != nil {
+		return true, err
+	}
+
+	scyllaClient, err := scc.getScyllaClient(ctx, sc, hosts)
+	if err != nil {
+		return true, err
+	}
+
+	klog.V(4).InfoS("Checking schema agreement", "ScyllaCluster", klog.KObj(sc))
+	hasSchemaAgreement, err := scyllaClient.HasSchemaAgreement(ctx)
+	if err != nil {
+		return true, fmt.Errorf("awaiting schema agreement: %w", err)
+	}
+
+	if !hasSchemaAgreement {
+		klog.V(4).InfoS("Schema is not agreed yet, will retry.", "ScyllaCluster", klog.KObj(sc))
+		return false, nil
+	}
+	klog.V(4).InfoS("Schema agreed", "ScyllaCluster", klog.KObj(sc))
+
+	// Snapshot system tables.
+
+	klog.V(4).InfoS("Backing up system keyspaces", "ScyllaCluster", klog.KObj(sc))
+	err = scc.backupKeyspaces(ctx, scyllaClient, hosts, systemKeyspaces, sc.Status.Upgrade.SystemSnapshotTag)
+	if err != nil {
+		return true, err
+	}
+	klog.V(4).InfoS("Backed up system keyspaces", "ScyllaCluster", klog.KObj(sc))
+
+	return true, nil
+}
+
+func (scc *Controller) afterUpgrade(ctx context.Context, sc *scyllav1.ScyllaCluster, services map[string]*corev1.Service) error {
+	klog.V(2).InfoS("Running post-upgrade hook", "ScyllaCluster", klog.KObj(sc))
+	defer klog.V(2).InfoS("Finished running post-upgrade hook", "ScyllaCluster", klog.KObj(sc))
+
+	hosts, err := controllerhelpers.GetRequiredScyllaHosts(sc, services)
+	if err != nil {
+		return err
+	}
+
+	scyllaClient, err := scc.getScyllaClient(ctx, sc, hosts)
+	if err != nil {
+		return err
+	}
+
+	// Clear system and data backup.
+
+	var wg sync.WaitGroup
+	errCh := make(chan error, len(hosts))
+	for i := range hosts {
+		host := hosts[i]
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			snapshots, err := scyllaClient.Snapshots(ctx, host)
+			if err != nil {
+				errCh <- fmt.Errorf("can't list snapshots on host %q: %w", host, err)
+				return
+			}
+
+			snapshotSet := setsutil.NewString(snapshots...)
+			for _, snapshotTag := range []string{
+				sc.Status.Upgrade.DataSnapshotTag,
```

Comment:
My thinking here is that you want to keep the snapshot for a rollback. As an admin I either wan't my cluster to rollout the new version or go back on all the nodes. Why were we removing it after the individual node upgrade?

Reply:
guess that may be unpractical with large clusters

Reply:
Reshape may be triggered during boot, and it may take couple of hours. This means that on large clusters upgrade may take even couple of days, so keeping snapshot for entire upgrade may be very inefficient. 
Maybe we should think about different snapshot policies for upgrades.

Reply:
ok, moved it back to post node upgrade